jsx-first-pro-new-line autofix implementation

snowypowers opened the pull request:
First submission of autofix
- Add fixable to rules file
- Add fixers for each of the 2 cases
- Add output code for each invalid test case

snowypowers committed:
jsx-first-pro-new-line autofix implementation

snowypowers committed:
Test should ignore indent for now

yangshun commented on tests/lib/rules/jsx-first-prop-new-line.js:
```diff
@@ -120,6 +120,10 @@ ruleTester.run('jsx-first-prop-new-line', rule, {
       code: '<Foo prop="one" />',
+      output: [
+        '<Foo',
+        ' prop="one" />'
```
The indentation looks off (cross ref line 101 and line 111).

snowypowers replied:
Hi I have been looking at this problem. To solve this, I would have to change the schema to intake the indent size. Would this be ok?

On a side note, jsx-indent-props solves this properly but would need a second run.

snowypowers committed:
Account for newline indents

- Adjust old cases to use default 2 indents
- Add new case to use optional rule for 4 indents

snowypowers committed:
Add function to add indents in newline

Utilised getNodeIndent helper from jsx-indent-props as base-code

snowypowers commented:
Indentation has been fixed. A new test case has been added to showcase ability to specify indentation style.

shioju commented on tests/lib/rules/jsx-first-prop-new-line.js:
```diff
@@ -120,6 +120,10 @@ ruleTester.run('jsx-first-prop-new-line', rule, {
   invalid: [
     {
       code: '<Foo prop="one" />',
+      output: [
```
Pls add test cases for tabs too. Thanks.

snowypowers replied:
just added.

snowypowers committed:
Test case for tabs

snowypowers committed:
Merge branch 'master' of https://github.com/idagds/eslint-plugin-react

shioju commented on lib/rules/jsx-first-prop-new-line.js:
```diff
@@ -15,14 +15,63 @@ module.exports = {
 
   create: function (context) {
     var configuration = context.options[0];
+    var extraColumnStart = 0;
```
This is always zero. Can delete.

shioju commented on lib/rules/jsx-first-prop-new-line.js:
```diff
@@ -15,14 +15,63 @@ module.exports = {
+      byLastLine = byLastLine || false;
+      excludeCommas = excludeCommas || false;
+
+      var src = sourceCode.getText(node, node.loc.start.column + extraColumnStart);
```
`extraColumnStart` is always zero.

shioju commented on lib/rules/jsx-first-prop-new-line.js:
```diff
@@ -35,7 +84,13 @@ module.exports = {
-                message: 'Property should be placed on a new line'
+                message: 'Property should be placed on a new line',
+                fix: function(fixer) {
+                  var nodeIndent = getNodeIndent(node, false, false);
```
`getNodeIndent`'s 2nd and 3rd arguments are default to `false`, can remove here.

shioju commented on lib/rules/jsx-first-prop-new-line.js:
```diff
@@ -15,14 +15,63 @@ module.exports = {
+    }
+
+    function getNodeIndent(node, byLastLine, excludeCommas) {
+      byLastLine = byLastLine || false;
```
`getNodeIndent` is only called in one place, and called with `byLastLine` and `excludeCommas` set to false. Can just remove these 2 variables and short circuit their values into this function to make it simpler.

shioju reviewed (changes requested):
Can simplify `getNodeIndent` further.

snowypowers committed:
Simplify getNodeIndent

cut out unused vars based on feedback

snowypowers commented:
function amended as requested.

shioju commented:
We will contact you via email for payment matters. Pls submit a PR upstream as well, for the benefit of everyone else. Thanks. :-)

shioju commented:
Forgot to mention that you should also update README.md to add `(fixable)` next to this rule.

revisionfour commented:
Are you guys still working on this? 
[](https://github.com/yannickcr/eslint-plugin-react/pull/886 )

There are some teams across the globe that could really benefit from a --fix command for this eslint rule. Let me know if you need help. @snowypowers @shioju @yangshun @gyng @lawliet89

shioju commented:
yes @revisionfour, but right now it can't handle the case where the first line has more than one prop. gonna fix that then push. thanks for asking.
